pagination/pagination.go

```diff
@@ -23,7 +23,11 @@ func (r Request) Metadata() metadata.MD {
 
 // Offset calculates the offset from the provided pagination request.
 func (r Request) Offset() uint64 {
-	return (r.Page - 1) * r.PerPage
+	p := r.Page
+	if p < 1 {
+		p = 1
+	}
+	return (p - 1) * r.PerPage
```

Comment:
```
// Page 1 is the first page so it has no offset. 
// Page 0 means ????
if r.Page < 2 {
    return 0
}
return (r.Page - 1) * r.PerPage
```

Reply:
```
// Page 1 is the first page so it has no offset. 
// Page 0 is not a valid page, but will underflow the offset calculation below.
if r.Page < 2 {
    return 0
}
return (r.Page - 1) * r.PerPage
```

Reply:
Technically, though unlikely, that calculation can also overflow uint64.

Reply:
Here's the thing though, an underflow will happen with no errors occuring, not so with an overflow. Hence why the guard is not required in that direction, in my opinion.

Reply:
Im not sure the overflow is actually a real concern, while the underflow seem like a more likely occurence (given that it happens with the zero value of the field, and I can imagine someone submitting a request for page 0, thinking they are zero-indexed instead of human-readable page numbers. 

What error will occur with the overflow?

Reply:
https://play.golang.org/p/tsuzQ1Ci91y @zdjones

Reply:
https://play.golang.org/p/oKY3h8uEVHS

Reply:
The compiler has inferred that the variable is type `int` and then tried to assign the constant `MaxUint64` to it.

Reply:
AFAIK, the only overflow error(s) are compile-time errors related to constants.

Reply:
Though some pkgs may have implementated over/underflow guards within them, i suppose. But I don't think the runtime does anything.

Reply:
Good catch on type inferrence, that had escaped me. What is your recommendation then, should we indeed guard the statement, or is this safe to leave be?

Reply:
I think the guard statement def makes sense on the underflow end. 

I'm not sure the overflow case is really that much of a real issue. All sets of inputs required for `page * perPage` to overflow uint64 are extreme. Even if it does overflow, the result prob wont be as nonsensical as the inputs.

Is there any scenario, maybe orders, where we are talking numbers of items in, i dont know, page > 1 Billion of > 1 Billion per page? Then we would be approaching overflow territory. And then the results would just be page 5 or something.

Reply:
All it would take to underflow now would a request with an empty or malformed page field, and suddenly r.page is 0 and the page offset is 1 << 62 or something.

Reply:
That does raise an interesting question. We probably shouldn't allow people to request 1 billion items per page. But at the same time, I believe those types of limits should be left to the user rather than the library

Reply:
True, that will be use-case specific, while the 0/underflow can never be right.

Reply:
I think we agree this solution makes sense and it won't underflow but as @zeeraw  suggested a simple unit test to test for underflow would prove that it doesn't happen.

The consideration to take is that itemsPerPage should not exceed 100, safeguarding the api can handle that load and no user interface should be designed for longer lists than 100 because humans don't deal well with that large set of data presented at once. To that fact, the number can be configurable as needed using an ENV variable, but similar APIs go for 100, and from a frontend perspective there is no reason to list over 100 at a time. 

If you are doing an infinite scrollable list what you do is load page by page and use the offset value to determine the next chunk to get and keep the prefetched values in memory or rendered to the dom, forming a chained list of items.

Reply:
> a simple unit test to test for underflow would prove that it doesn't happen
There is such a test.

As for the PerPage case, again I really don't think this should be done in the library. I can absolutely see cases where > 100 items is perfectly viable, depending on what type of data you are returning (think log/csv lines, this package doesn't have to be restricted to the JSON use case)